src/msg/xio/XioMessenger.cc

```diff
@@ -388,6 +391,30 @@ int XioMessenger::pool_hint(uint32_t dsize) {
 				   XMSG_MEMPOOL_QUANTUM);
 }
 
+void XioMessenger::learned_addr(const entity_addr_t &peer_addr_for_me)
+{
+  // be careful here: multiple threads may block here, and readers of
+  // my_inst.addr do NOT hold any lock.
+
+  // this always goes from true -> false under the protection of the
+  // mutex.  if it is already false, we need not retake the mutex at
+  // all.
+  if (!need_addr)
+    return;
+
+  sh_mtx.Lock();
```

Comment:
if the performance of XioMessage could be hurt by the lock contention here due to frequently called `on_new_session`. we might want to use a `ceph::atomic_t` and `compare_and_swap()` here.